app/controllers/registrations_controller.rb

```diff
@@ -34,6 +34,53 @@ class RegistrationsController < ApplicationController
     redirect_to competition_path(competition_from_params) if competition_from_params.use_wca_registration?
   end
 
+  before_action :check_errors_import_registration!, only: %i[do_import]
+  private def check_errors_import_registration!
+    @competition = competition_from_params
+    file = params[:csv_registration_file]
+
+    competition_events = @competition.competition_events
+    required_event_columns = competition_events.map(&:event_id)
+    required_other_columns = ["status", "name", "country", "wca id", "birth date", "gender", "email"]
+
+    headers = CSV.read(file.path).first.compact.map(&:downcase)
+    missing_headers = (required_other_columns + required_event_columns) - headers
+    return render status: :unprocessable_entity, json: { error: I18n.t("registrations.import.errors.missing_columns", columns: missing_headers.join(", ")) } if missing_headers.any?
+
+    filtered_csv_rows = CSV.read(
```

Comment:
Why are you doing `CSV.read` here again? Wouldn't it be enough to load the CSV once (with all proper settings and everything) and then perform all necessary checks on the already loaded CSV?